src/Chainweb/OpenAPIValidation.hs

```diff
@@ -0,0 +1,70 @@
+{-# language TupleSections #-}
+{-# language ScopedTypeVariables #-}
+{-# language FlexibleContexts #-}
+{-# language OverloadedStrings #-}
+
+module Chainweb.OpenAPIValidation (mkValidationMiddleware) where
+
+import Control.Monad
+import Data.Aeson
+import qualified Data.ByteString.Char8 as BS8
+import qualified Data.ByteString.Lazy as BL
+import Data.Foldable
+import qualified Data.HashSet as HS
+import Data.IORef
+import Data.LogMessage (LogFunctionText)
+import qualified Data.Text.Encoding as T
+import qualified Data.Yaml as Yaml
+import qualified Network.HTTP.Client as HTTP
+import Network.Wai
+import qualified Network.Wai.Middleware.Validation as WV
+import System.LogLevel
+
+import Chainweb.Logger
+import Chainweb.Time
+import Chainweb.Utils
+import Chainweb.Version
+
+mkValidationMiddleware :: Logger logger => logger -> ChainwebVersion -> HTTP.Manager -> IO Middleware
+mkValidationMiddleware logger v mgr = do
+    (chainwebSpec, pactSpec) <- fetchOpenApiSpecs
+    apiCoverageRef <- newIORef $ WV.initialCoverageMap [chainwebSpec, pactSpec]
+    apiCoverageLogTimeRef <- newIORef =<< getCurrentTimeIntegral
+    return $
+        WV.mkValidator apiCoverageRef (WV.Log logValidationFailure (logApiCoverage apiCoverageLogTimeRef)) $ \path -> asum
+            [ case BS8.split '/' path of
+                ("" : "chainweb" : "0.0" : rawVersion : "chain" : rawChainId : "pact" : "api" : "v1" : rest) -> do
+                    findPact pactSpec rawVersion rawChainId rest
+                ("" : "chainweb" : "0.0" : rawVersion : "chain" : rawChainId : "pact" : rest) -> do
+                    findPact pactSpec rawVersion rawChainId rest
+                _ -> Nothing
+            , (,chainwebSpec) <$> BS8.stripPrefix (T.encodeUtf8 $ "/chainweb/0.0/" <> chainwebVersionToText v) path
+            , Just (path,chainwebSpec)
+            ]
+    where
+    findPact pactSpec rawVersion rawChainId rest = do
+        reqVersion <- chainwebVersionFromText (T.decodeUtf8 rawVersion)
+        guard (reqVersion == v)
+        reqChainId <- chainIdFromText (T.decodeUtf8 rawChainId)
+        guard (HS.member reqChainId (chainIds v))
+        return (BS8.intercalate "/" ("":rest), pactSpec)
+
+    fetchOpenApiSpecs = do
+        let chainwebUri = "https://raw.githubusercontent.com/kadena-io/chainweb-openapi/validation-fixes-3/chainweb.openapi.yaml"
```

Comment:
It seems a little iffy to me to point at a non-permanent link here. What exactly is the intention here? If I use a chainweb binary that was compiled before the latest update to the `validation-fixes-3` branch, why is it an error that my binary doesn't conform to the spec there? Why is it the `validation-fixes-3` branch anyway?

In any case, the reasoning here needs explanation.

Reply:
My explanation is, we need review on the validation-fixes-3 branch to merge it into master :D

Reply:
:D that's definitely an improvement! Still though, when we merge new changes into `chainweb-openapi`'s `master` branch,  existing node binaries will start validating differently. Does that make sense? Especially in the context of tests, is it OK that if I go back and execute the tests of a `chainweb-node` commit from a year ago, they will fail since the API spec living in the `master` branch will have changed?

Reply:
It's not great, I would prefer that the spec be inside of this repo.